File: julia/Weibull.jl
using SpecialFunctions
using LinearAlgebra

#******************************************************************************
function med(c, b)
    return c*gamma(1.e0 + 1.e0/b)
end
#******************************************************************************

#******************************************************************************
function desv(c, b)

    mu = med(c, b)
    dev = (c^2)*gamma(1.e0+2.e0/b) - mu^2
    return sqrt(dev)
end
#******************************************************************************


#******************************************************************************
function get_b_c(mu, sig, tol = 1e-11, maxIt=100000)

    function solv(A, x, F)
        x[1] = F[1]/A[1]
        x[2] = F[2]/A[2]
        x[3] = (F[3] - A[4]*x[2])/A[3]
    end


    function vetorF(F, sig, mu)
        F[1] = mu
        F[2] = mu^2 + sig^2
        F[3] = 0.e0
    end

    function matvec(A, x)
        y1 = A[1]*x[1]
        y2 = A[2]*x[2]
        y3 = A[3]*x[3] + A[4]*x[2]
        return [y1, y2, y3]
    end

    function Matriz(A, x)
# linha 1
         A[1] = x[3]*gamma(x[3])   # A11
# linha 2
         A[2] = (x[1]^2)*gamma(x[2]) # A22
# linha 3
        A[3] =  2.e0               # A33
        A[4] = -1.e0               # A23
    end

# ... inicializado arranjos
    A   = zeros(Float64, 4)
    x   = zeros(Float64, 3)
    F   = zeros(Float64, 3)
    alf = 0.5e0
# ... chute inicial
    x[1] = mu        # c
    x[2] = sig/mu    # y
    x[3] = 2.e00*x[2]  # x
# ...
    vetorF(F, sig, mu)

# ...
    global res, j
    j = 0
    while j < maxIt
# ...
        x0 = copy(x)
# ...
        Matriz(A, x)
# ... residuo
        R = F - matvec(A ,x)
# ... |R*R|
        res = norm(R, 2)
        if res < tol
            break
        end
        j+=1
# ...
        solv(A, x, F)
# ... relaxamento
        x = alf*x + (1.e0 - alf)*x0
    end
# .............................................................................

    println("**********************************************************")
    println("Res final ", res , " para it ", j)
    println("**********************************************************")

# ...
    c = x[1]
    b = 2.0/x[2]

    return b, c
end
#******************************************************************************

#******************************************************************************
function main()

    mu  = 99884.95e0
    sig = 255.84e0

    @timev b, c = get_b_c(mu, sig, 1.e-15, 1500000)

    println("**********************************************************")
    println("c = ", c)
    println("b = ", b)
    println("**********************************************************")

    mui  = med(c, b)
    sigi = desv(c, b)

    println("**********************************************************")
    println("mu    = ", mu)
    println("mui   = ", mui)
    println("|res| = ", abs(mu - mui))
    println()
    println("sig   = ", sig)
    println("sig   = ", sigi)
    println("|res| = ", abs(sig - sigi))
    println("**********************************************************")

end
#******************************************************************************

main()

File: python/Weibull.py
import numpy as np
import math as mt

# **********************************************************
def media(c,b):
    '''
    Media apartir dos paramentros c e b
    '''
    return c*mt.gamma(1.e0+1.e0/b)
# **********************************************************

# **********************************************************
def desv(c,b):
    '''
    desvio padrao apartir dos paramentros c e b
    '''
    mu = media(c,b)

    dev = mt.sqrt((c**2)*mt.gamma(1.e0+2.e0/b) - mu**2)  

    return dev
# **********************************************************

# **********************************************************
def get_b_c(mu,sig,tol=1.e-11,maxIt=10000):
    '''
    ***********************************************************
    data criacao:     09/01/2020
    data modificacao: 00/00/0000
    -----------------------------------------------------------
    get_b_c: obtem o b e c dada uma media e um desvio padrao
    -----------------------------------------------------------
    Entrada:
    -----------------------------------------------------------
    mu    - media
    sig   - desvio padrao
    maxIt - numero maximo de iteracoes
    tol   - tolerancia desejada
    -----------------------------------------------------------
    Saida:
    -----------------------------------------------------------
    b e c - 
    -----------------------------------------------------------
    OBS:
    -----------------------------------------------------------
    ***********************************************************
    '''
    def solv(A,x,F):
        '''
        solver do Ax=F
        '''
        x[0] = F[0]/A[0]
        x[1] = F[1]/A[1]
        x[2] = (F[2] - A[3]*x[1])/A[2]

    def MatrizA(A,X):
        '''
        atualizacao da matriz A(x)
        '''
        c = X[0]
        y = X[1]
        x = X[2]

# linha 1
        A[0] = x*mt.gamma(x)   # A11
# linha 2
        A[1] = c*c*mt.gamma(y) # A22
# linha 3  
        A[2] =  2.e0            # A33
        A[3] = -1.e0            # A23

    def VetorF(F,sig,mu):
        '''
        calculo do vetor F
        '''
        F[0] = mu
        F[1] = sig**2 + mu**2
        F[2] = 0.0

    def matvec(A,x):
        '''
        operacao matriz vetor
        '''
        y = np.zeros(3,dtype=np.float64)

        y[0] = A[0]*x[0]
        y[1] = A[1]*x[1]
        y[2] = A[2]*x[2] + A[3]*x[1] 

        return y

# ...
    A = np.zeros(4,dtype=np.float64)
    x = np.zeros(3,dtype=np.float64)
    F = np.zeros(3,dtype=np.float64)

    alf   = 0.5e0
# ... chute inicial
    x[0] = mu     # c
    x[1] = sig/mu # y
    x[2] = 2.e0*x[1] # x

    VetorF(F,sig,mu)


# ... loop
    for i in range(1,maxIt+1):
        MatrizA(A,x)        
# ... residuo
        R = F - matvec(A,x)
# ... ||R||    
        Res = np.linalg.norm(R)
        if  Res < tol:       
            break
# ...
        x0 = np.copy(x)
# ... x = (A^-1)*F
        solv(A,x,F)
# ... update x
        x = alf*x + (1.e0 - alf)*x0 

    print('\n***************************************')
    print('Res final {0} para it {1}'.format(Res,i))
    print('***************************************')

    c = x[0]
    b = 2.e0/x[1]
    
    return b,c
# **********************************************************




